### ISSUE TITLE
[Returned for Revisions] SDL 0241 - Commerce App Service

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#776]
Hello SDL community,

The review of "SDL 0241 - Commerce App Service" begins now and runs through July 16, 2019.  The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0241-Commerce-App-Service.md

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/776

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by joeljfischer** [comment]
I think a service like this could be very useful.

1\. A better name for this service is probably "POINTS_OF_INTEREST" because it can relate to things other than commerce (e.g. hiking trails). Then other structures like `SearchServiceManifest` should be renamed to `POIServiceManifest`.

2\. I would appreciate if the inconsistent spacing could be cleaned up. Sometimes there's the correct 4 space indentation, othertimes 8 or other amounts.

3\. I don't think this proposal is comprehensive enough. It should have several examples that describe the usage. e.g. How might this be used with a coffee house app, a hiking app, and a mapping app like Google Maps? Ideally there would be mockup UI to describe how it would be used within an embedded navigation app or something like that.

4\. I think it should be defined how many `SearchServiceData` `POIs` are permitted in the array (e.g. a maximum of 100), the order (it should probably be closest to current GPS coordinates and moving outward), and that these POIs relate to the user's current GPS coordinates (e.g. rename the `POIs` parameter to `currentLocationPOIs`).

5\.  I have some questions about the `SearchServiceData` `softButtons` parameter.

(a) The `POI` `softButtons` only supports one `SoftButton`, was that intended?

(b) It should be defined that if the `POI` `softButtons` support `STEAL_FOCUS `, then they should open the app service's originating app. This may require additional implementation steps in Core / HMI to properly do this. 

(c) I'm assuming that the soft buttons would have to apply to every POI displayed. I'm unsure if this is the right way to go.

(d) How would the `onButtonPressed` notification work with this?

6\. I'm not sure how the `image` in the `SearchServiceData` would be used. It would seem better to have the POI support the `image` field, right? What is one image going to do for all the `POI` data? Where would it even show up?

7\. I would recommend additional changes to `POI`. 

(a) `Name` should be `name`. 

(b) Instead of the `softButtons` on `searchServiceData`, which I'm assuming would be used for things like calling the POI, navigating to the POI, or opening the app to a template with POI data, I'd ask others to think about a slightly different structure.

(7.b.i) We allow only one soft button, an "Open in <Producer App>" button to open the POI in the producer app.
(7.b.ii) Add additional POI data akin to `SendLocation`, like `address`, `phoneNumber`, and `locationDescription`. This would allow the consumer app to provide buttons like "call", "navigate to", and additional textual information.

8\. I'm unsure of the utility of `PerformAppServiceInteraction Uri`. If we have the soft buttons sending the correct button presses and events (see 5.d) then this wouldn't be necessary. If we wanted the "Open in <Producer App>" button (see 7.b.i) to be handled via a URI instead of `onButtonPress`, then we'd have to make it required somehow. This would be a rather awkward implementation.

9\. I have notes on the `POISearch` RPC:

(a). Why did you go with the `keywords` parameter instead of a simpler solution like:

```xml
<param name="searchTerm" type="String" mandatory="true" />
```

Users don't usually add keywords, they search for terms. The `keywords` would usually just be a split up search term, right?

(b). The function ID should be `POISearchID`.

10\. The downside should be mitigated in the proposed solution itself. Core would need to store multiple active services and HMIs would need to be able to enable disable multiple app services in their UI. Sending an RPC like `POISearch` would need to be distributed to multiple app service producers, and then multiple responses would be returned. This is a huge change that is not currently supported by SDL. Therefore, I think this "downside" is a killer unless we can mitigate it.

(a). We could not support _any_ active services for this service type and instead add additional capabilities for getting service data from inactive services. This doesn't solve our `POISearch` problem.

(b). We could change the `POISearch` request / response pair into notification pair of some sort.

**Comment by joeygrover** [comment]
I agree with most of Joel's observations. I think more detail in the proposal would help others understand what the service should do and how it could be used from both a producer and consumer perspective. 

7\. I notice `GPSData` is used, but that is the structure that is sent from vehicle data requests, I believe it should be [`LocationDetails` type](https://github.com/smartdevicelink/rpc_spec#locationdetails). If that is the case, it already has the `name` and `image` field that could be reused. At that point I believe we would have to weight the usefulness of a new struct that wraps LocationDetails and if there are any params that would be included, maybe the soft button objects as pointed out by Joel?

8\. I think we're getting into a point of discussion made on my previously requested revision to app services. Do we send the structured URI scheme in an ordered array that the OEM can then display and send the constructed URI? Or keep the general soft button methodology? The provided method is hard to follow though. It seems like we're trying to define schema that these apps will have to follow while also telling them to use soft buttons. 

10\. This is a big issue that can't be swept under the rug. It will have a big impact on how core handles app services and needs to be thought out on precisely how it would work. Instead of multiple active services, could the services be presented to the user for a choice? Essentially select the service they want to use for POIs, which would end up being sort of a "category" selection?

11\. Another item here is defining how a POI provider would work. Are we assuming that a POI provider would only supply one "type" of POIs (coffee shops, parks, etc)? Or would something like a popular POI review app supply multiple types of POIs? Would there be collisions? If so, how would they be resolved? This relates to item number 10 as well I believe. If only one active service is active at a time this seems to be resolved.

12\. `POISearch` response doesn't actually contain any POIs. Is it assumed that this would simply request the app update their service data? This relates to number 9, wouldn't this be better as a defined schema the app should follow? Relating to number 10, if we have multiple active services this RPC would have to be sent to all of those services, and how would the response be calculated by core?

**Comment by theresalech** [comment]
The Steering Committee voted to defer this proposal, keeping it in review until our next meeting, to allow the author time to respond to the comments on the review issue.

**Comment by khburdette** [comment]
1. This makes sense

2. I will clean it up

3. I will add examples

4. This also makes sense

5. My original thought was that we only needed to send 1 soft button to be used for all POIs. I agree with your feedback and think it makes more sense to allow every POI to have its own soft button. This would make `onButtonPressed` cleaner and remove the need for the `PerformAppServiceInteraction` Uri (point 8)

6. The though for the image was to be able to show an image on a POI marker on a map like a logo. Yes, it makes sense to be part of the POI instead. If we use `LocationDetails`, we don't need the image field anymore

7. I like the idea of using `LocationDetails`. Instead of a new struct, would it make sense to add soft buttons to `LocationDetails`? Or should I keep the new struct and have it include `LocationDetails` and soft buttons?

8. When using 1 soft button for all POIs, the app would need the `onButtonPressed` to know if the soft button was pressed as well as the location that the soft button was pressed at. The Uri was intended to send that location information. If we go with a soft button for every POI (see my comment on 5), then this becomes unnecessary

9. You're right, I'll make this change

10. This app service is the first where we're proposing multiple active services, so, you're right, it's a big issue, and it might set precedence. Couple thoughts:

  (a). When an RPC can go to multiple apps, should we be able to specify a service to send it to?

  (b). Possibly be fixed by making `POISearch` a notification, but this probably won't scale

11. Good point. If the consumer has to send to specific services, it puts the burden on the consumer to prioritize one service over another in case there's conflicts.

12. Do you think it would be better to put the POI list as a response to `POISearch` instead of in the service data? When thinking about it, if you have more than 1 consumer requesting POIs, how do they know the service data if for them ... so it definitely needs to be cleaned

**Comment by joeygrover** [comment]
Overall I think there is a decent amount that is agreed upon to send back for revisions. The biggest issue with the proposal is still outstanding though. I'm not sure there is a clear answer on how to handle multiple "active" services at a time, or a workaround that lets one active service still work.

5\. Just to clarify, are we agreeing to an array of softbuttons? Or still just a single softbutton?
6\. & 7\. The more I thought about it the more I do think we should have a `POI` struct that contains the `LocationDetails`, `softbuttons`, and an `icon`. The `icon` would be the image used on a map, while the image in the` locationDetails` could be more like a photograph of the location, or something more detailed vs a small icon. 
10\. a) Right now there is no way for an RPC to be sent to a specific service; the exception to this rule is the `PerformAppServiceInteraction`. Enabling a way to send the RPC to specific destinations would be a big alteration.
b) I think we still run into an issue here using a notification, it just doesn't depend on a single response. We would get multiple Requests from the app after that point, or updates form services.
11\. Right, so I'm not sure the best way to handle this. It has implications related to solving the multiple active services.
12\. This will also relate to the if of multiple active services is included. I think flushing out some use cases is the next step of us to determine how we can move forward in that regard.

**Comment by joeljfischer** [comment]
5. We could also add a new `ButtonName` enum value and require that apps be able to open a POI in their app. `SoftButtons` become difficult to route `onButtonPress` and `onButtonEvent` correctly. We'll at least need to work out how to do that.

**Comment by theresalech** [comment]
The Steering Committee voted to return this proposal for revisions, so that the author can update the proposal as summarized in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/776#issuecomment-514366737).  It was noted that adding use cases will help the Steering Committee to understand the larger issue the proposal is trying to solve, and if the author has ideas on how to tackle multiple active services, he can include those in the revisions as well.

**Comment by theresalech** [comment]
Closing as inactive.  The issue will remain in a `returned for revisions` status and unlocked so the author or any other interested party can notify the Project Maintainer when it is ready to be revisited.
